Looking for feedback: add ability to evaluate get_status() methods in gcode_macro

KevinOConnor opened the pull request:
The menu system has the ability to query the internal state of the machine via `get_status()` methods.  It may also be useful to be able to do this from a gcode_macro.  This change adds this ability via a new builtin "status" variable in gcode_macro.

For example:
```
[gcode_macro MY_MACRO]
gcode:
    ECHO FAN_SPEED={status[fan][speed]}
    ECHO EXTRUDE_FACTOR={status[gcode][extrude_factor]}
```

I'm not sure how useful this is without some basic math evaluation.  It would also likely need lots more "get_status()" methods to be added, but that's not really a bad thing.

@mcmatrix , @Arksine , @FHeilmann - FYI, this may impact code you're working on.

Thoughts?
-Kevin

alecplumb commented:
This looks great. Thank you for writing it so I don't have to. 

Two suggestions:
1) Add an `enabled` option to `[gcode_macro]`, that can evaluate a status variable to decide whether the macro should be run. For example, you may only want the macro to be executable while not printing.

2) Evaluate each line of the gcode macro in sequence, after executing the previous line.

Arksine commented:
I can see this being useful for creating macros for some of the more obscure gcodes not natively supported by klipper.  For example, M31:

```
[gcode_macro M31]
gcode:
 M118 {status[toolhead][printing_time]}
```

I'm sure as more modules implement get_status() people could come up with some more creative macros using it.

FHeilmann commented:
I have a little work done on a `variable_store` extra that allows users to define, set and use global parameters in gcode macros. See here: https://github.com/FHeilmann/klipper/blob/pr_variable_store/klippy/extras/variable_store.py
I made a small change to `gcode_macro` to query the variables stored in this extra in addition to its own to make this work.
Your work sounds like it could easily integrate with that extra, or integrate the concept of user defined variables into the gcode_macro extra altogether. I'll leave that decision up to you.
As @Arksine said this would allow klipper to 'implement' many gcodes by using macros. My current use-case for this were Firmware Retraction settings where the distances, speeds and zhops were adjustable using extended gcodes.

As for math expression evaluation there are a couple of libraries we could use such as pyparsing as described here: https://stackoverflow.com/a/2371789

simonkuehling commented:
I really appreciate the aim to get more runtime data available in gcode macros, nice!

Your proposed array-like implementation through a "status" variable is very convenient.
Math evaluation would be super powerful to extend on that - but I would suggest to also look into implementing basic logic evaluation (if/then/else) at the same time in order to act differently depending on the current runtime status.
The complexity of routines one could build with this would be super handy for active toolchanger setups and such - checking safety sensors ("tool attached?") etc. during macros for example.

If you look at what is possible just by gcode programming (routines, logic, math) in the CNC mill/lathe world, that is pretty amazing!

mcmatrix commented:
Thank you. It looks interesting.
I like that more "get_status()" methods will be added or extended.
Menu system will definitely take benefit from it.

I'll investigate can I use and integrate the same StatusWrapper in the menu code
and switch from dot notation to this or even get rid of indexes and use names.

I also like the idea of simple math expression parser (maybe even some simplified predefined function calls, like menu "transform" to change or transform value) all in one engine.

mcmatrix commented:
The other option is to integrate the same template engine "**Jinja**" that Octoprint is using in his gcode scripts. Anyway, this is a slippery way and is definitely "programming in the config file".
Maybe there is a way to strip/disable some functionalities from that template engine and allow only what you think is safe to do in the config?

mcmatrix commented:
I played a little bit with that named menu parameters idea.

This is how it's at the moment.
Input element takes `gcode speed_factor` parameters which is float.
It shows it in the menu as percentage formatting and will interpolate input value (index 0 default)
from `0.0 - 2.0 to 0 - 200` (index 1) and will use it in gcode script.
```ini
[menu __tune __speed]
type: input
name: "Speed: {1:3d}%"
parameter: gcode.speed_factor
transform:
    map(0,2,0,200)
input_min: 0
input_max: 2
input_step: 0.01
realtime: true
gcode: M220 S{1:d}
```

The same by using named parameter approach.
```ini
[menu __tune __speed]
type: input
name: "Speed: {__input:3d}%"
input: status[gcode[speed_factor]]
transform_speed: __input.map(0,2,0,200)
input_min: 0
input_max: 2
input_step: 0.01
realtime: true
gcode: M220 S{transform[speed]:d}
```

Fan control menu element
It's using `fan speed` parameter as input (index 0 default).
Will transform it to `ON OFF` string (index 1) and to
integer `0 - 255` (index 2)

```ini
[menu __control __fanonoff]
type: input
enable: fan.is_enabled
name: Fan {1:3s}
parameter: fan.speed
transform:
    choose('OFF','ON')
    choose(0,255)
input_min: 0
input_max: 1
input_step: 1
gcode: M106 S{2:d}
```

The same by using named approach.
```ini
[menu __control __fanonoff]
type: input
enable: fan[is_enabled]
name: Fan {transform[onoff]:3s}
input: fan[speed]
transform_onoff: __input.choose('OFF','ON')
transform_gcode: __input.choose(0,255)
input_min: 0
input_max: 1
input_step: 1
gcode: M106 S{transform[gcode]:d}
```

The named approach looks much more clear.

Are you going forward with this `StatusWrapper` development?
I think it should be Global so that all functionalities will use the same parameter subsystem.

KevinOConnor commented:
> The named approach looks much more clear.

Yes, I agree and I think it would be worthwhile to do.

One minor recommendation would be to try and use less special cases - so perhaps the __input can just be replaced with "input", "transform[gcode]" could be replaced with "transform_gcode", and transformation syntax could be something like "choose(input, 'OFF','ON')".

> Are you planning to go forward with this StatusWrapper development?

Yes, but I also wanted to prototype a template engine (eg, Jinja2) just to see how that looks first.  If it helps, you can assume the StatusWrapper will be available as I think we'll need that regardless of the template results.

-Kevin

mcmatrix commented:
> Yes, but I also wanted to prototype a template engine (eg, Jinja2) just to see how that looks first. If it helps, you can assume the StatusWrapper will be available as I think we'll need that regardless of the template results.

That's great.
Btw, Is it okay for the performance to create a new StatusWrapper instance on each menu refresh?

KevinOConnor commented:
I got a chance to take a look at the Jinja2 template engine.  If nothing else, it's super simple to integrate with Klipper (basically just commit 393e1ae9).  This uses '{{ ... }}' for blocks and '{ ... }' for expressions.

Simple expressions work as expected - for example:
```
[gcode_macro my_test]
gcode:
  echo status={status.gcode.move_xpos + status.gcode.move_ypos - 17}
```

Blocks also work:
```
[gcode_macro my_test_block]
gcode:
  {{ for i in range(3) }}
  echo i={i} a={A} status={status.gcode.move_xpos}
  {{ endfor }}
```

I'm not sure the templating language is a great fit for Klipper (it's heavily html focused), but the super simple integration is a big plus.

Comments?
-Kevin

FHeilmann commented:
Looks very nice!
If it's simple enough to integrate and works as expected, why not use it! I feel the most 'thinking' has to go into a well-defined methodology when get_status gets implemented for the many extras and parts of klippy.
- Florian

simonkuehling commented:
I second that - looks clean and well worth rolling out!
Looking forward to using it in new machine developments so much :-)

KevinOConnor commented:
FYI - I made some additional updates to this test branch.  I changed the block delimiters back to Jinja2's default of "{% ... %}" (but I've kept the simpler expression eval of "{ ... }").  I've also updated the code to implement the templates on homing_override and probe activate/deactive_gcode blocks.

-Kevin

mcmatrix commented:
Nice work.
I was thinking that I could use the same engine for the menu.
There's no point to implement an additional `ast` parser for the menu.

Is it possible to add more variables and functions when using `load_template` and `TemplateWrapper`?
For example, `input` type menu has an additional variable `input` that I want to add before rendering.
The menu also has helper functions to transform value to something else. It would be nice to use these functions too in this template engine.

KevinOConnor commented:
Yes - I have template loading working locally (see code below).  The jinja2 syntax is pretty goofy though, so I was going to hold off on merging the loading capability until there is a user for it.

FWIW, I've thought of a few things that would be useful tasks post-merge:
1 - deprecate gcode_macro default_parameter_X capability.  It's now possible to accomplish the same thing via `param.X|default("abc")`.
2 - add in an "action" / obj.get_actions() capability similar to the "status" / obj.get_status() mechanism.  Particularly useful actions would be action_emergency_stop(), action_respond_info(), and action_respond_error().  Since menu and filament sensor templates would be evaluated at the time of the button event, this mechanism would allow the templates to take important actions more timely.
3 - it would be useful to have a SET_GCODE_MACRO_VARIABLE type of command so that macros could save/restore values between invocations.  (As @FHeilmann points out.)
4 - all this stuff needs to be documented.

-Kevin

```diff
--- a/klippy/extras/gcode_macro.py
+++ b/klippy/extras/gcode_macro.py
@@ -57,15 +57,32 @@ class TemplateWrapper:
     def run_gcode_from_command(self, context=None):
         self.gcode.run_script_from_command(self.render(context))
 
+# Template caching class
+class TemplateLoader(jinja2.BaseLoader):
+    def __init__(self):
+        self.cache = {}
+    def add_template(self, name, template):
+        self.cache[name] = template
+    def get_source(self, environment, template):
+        raise jinja2.TemplateNotFound(template)
+    def load(self, environment, name, globals=None):
+        template = self.cache.get(name)
+        if template is None:
+            raise jinja2.TemplateNotFound(template)
+        return template
+
 # Main gcode macro template tracking
 class PrinterGCodeMacro:
     def __init__(self, config):
         self.printer = config.get_printer()
-        self.env = jinja2.Environment('{%', '%}', '{', '}')
+        self.loader = TemplateLoader()
+        self.env = jinja2.Environment('{%', '%}', '{', '}', loader=self.loader)
     def load_template(self, config, option):
         name = "%s:%s" % (config.get_name(), option)
         script = config.get(option, '')
-        return TemplateWrapper(self.printer, self.env, name, script)
+        tw = TemplateWrapper(self.printer, self.env, name, script)
+        self.loader.add_template(name, tw.template)
+        return tw
 
 def load_config(config):
     return PrinterGCodeMacro(config)
```

KevinOConnor commented:
@mcmatrix - Ah, I think I answered a different question than the one you were asking. 

Yes, the `context` parameter of TemplateWrapper.render() allows one to pass in additional variables and functions.

As for defining functions - the Jinja2 syntax has a "template loading" mechanism (which is what I thought you were originally asking about) that we could use.  Alas, the syntax for it is pretty goofy.  Another option, as you indicated, would be to pass in some common functions via the context parameter.  Finally, another option would be to define per printer object "action" methods that we could globally expose to the templates.

-Kevin

KevinOConnor commented:
FYI, I've merged this branch into the master branch.

-Kevin
